### PR TITLE
Issue 15118: Add a cache of versioned KeyValueSchemaImpl

### LINKED ISSUE #15118: Cache key and value schema versions in KeyValueSchemaImpl
**Is your enhancement request related to a problem? Please describe.**
While doing performance profiling, @eolivelli found that `KeyValue::decode` is called twice each time `KeyValueSchemaImpl::atSchemaVersion` is called (once for the key, the other time for the value). 
This can be for each message you receive from an AUTO_CONSUME consumer for instance.
Eg. in the following code the decoding happens for each call to `getReaderSchema`
```java
    @Test
    public void test() throws Exception {
        String topic = "my-property/my-ns/schema-test" + UUID.randomUUID();

        Schema<KeyValue<V1Data, V1Data>> pojoSchema = Schema.KeyValue(
                Schema.AVRO(V1Data.class),
                Schema.AVRO(V1Data.class),
                KeyValueEncodingType.INLINE);

        try (Producer<KeyValue<V1Data, V1Data>> p = pulsarClient.newProducer(pojoSchema)
                .topic(topic)
                .create();
             Consumer<GenericRecord> c0 = pulsarClient.newConsumer(Schema.AUTO_CONSUME())
                     .topic(topic)
                     .subscriptionName("sub0")
                     .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest)
                     .subscribe();
        ) {

            p.send(new KeyValue<>(new V1Data(1), new V1Data(2)));
            p.send(new KeyValue<>(new V1Data(1), new V1Data(2)));

            c0.receive().getReaderSchema();
            c0.receive().getReaderSchema();
        }
    }
```
So it would be beneficial to cache the result of [the key and value atSchemaVersion calls](https://github.com/apache/pulsar/blob/18f2f4a9c1dab7eec5c7c9590b76aca17ee1ead8/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaImpl.java#L285) to spare some CPU cycles.

**Describe the solution you'd like**
Add `ConcurrentMap<SchemaVersion, Schema<?>> keySchemaMap` and `ConcurrentMap<SchemaVersion, Schema<?>> valueSchemaMap` in `KeyValueSchemaImpl` and get/put from/to cache in `KeyValueSchemaImpl::atSchemaVersion`

**Describe alternatives you've considered**
none

**Additional context**
image

### PR DESCRIPTION
**Pull request by cbornet**
Fixes #15118

### Modifications

Added `schemaMap` and to cache results of `atSchemaVersion` in `KeyValueSchemaImpl `

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
* Run `KeyValueSchemaTest:: testKeyValueSchemaCache `

### Does this pull request potentially affect one of the following parts:

no

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

File: pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaTest.java [hunk 1]
```diff
@@ -400,4 +404,18 @@ public void testKeyValueSchemaSeparatedEncoding() {
...
+
+    @Test
+    public void testSchemaCaches() {
+        Schema<Foo> keySchema = spy(Schema.AVRO(Foo.class));
+        Schema<Foo> valueSchema = spy(Schema.AVRO(Foo.class));
+        KeyValueSchemaImpl<Foo, Foo> keyValueSchema = (KeyValueSchemaImpl<Foo,Foo>)
+                KeyValueSchemaImpl.of(keySchema, valueSchema, KeyValueEncodingType.SEPARATED);
+
+        keyValueSchema.atSchemaVersion(new byte[0]);
+        keyValueSchema.atSchemaVersion(new byte[0]);
```
Comment (eolivelli) [inline pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaTest.java:416]: can we assert that we are returning the same object ?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaImpl.java [hunk 2]
```diff
@@ -292,14 +298,27 @@ public String toString() {
...
 
+    private Schema<?> getSchemaAtVersion(Schema<?> schema,
+                                         Map<SchemaVersion, Schema<?>> schemaMap,
+                                         byte[] schemaVersion) {
+        if (!(schema instanceof AbstractSchema)) {
+            return schema;
+        }
+        Objects.requireNonNull(schemaVersion);
+        SchemaVersion cacheSchemaVersion = BytesSchemaVersion.of(schemaVersion);
+        schemaMap.computeIfAbsent(
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaImpl.java:315]: what about "return schemaMap.computeIfAbsent(" ...
this way we save a "get"

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaTest.java (hunk 1 above)
Comment (cbornet) [reply pulsar-client/src/test/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaTest.java:416]: Sure

**Review by michaeljmarshall — COMMENTED** [review]
@cbornet - can you explain a bit more how your solution prevents decoding the schema twice? It seems to me that it is decoded twice because of the way that the `KeyValueSchemaImpl` class defines the `SchemaInfoProvider` for both the `keySchema` and the `valueSchema`:

https://github.com/apache/pulsar/blob/c7c00db0f7cdfa36d492e66f645263e7973f63a4/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaImpl.java#L246-L250

https://github.com/apache/pulsar/blob/c7c00db0f7cdfa36d492e66f645263e7973f63a4/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/KeyValueSchemaImpl.java#L265-L269

Note that both `SchemaInforProvider` instances close over the same `schemaInfoProvider`, so `KeyValueSchemaInfo.decodeKeyValueSchemaInfo(si).getKey()` and `KeyValueSchemaInfo.decodeKeyValueSchemaInfo(si).getValue()` both decode the whole schema and then throw away the value or the key, respectively. I think we might need to look at optimizing the way the `SchemaInfoProvider` class is implemented.

**Comment by cbornet** [comment]
The fix is not to prevent from decoding twice. It is to prevent from decoding each time we call `KeyValueSchemaImpl:: atSchemaVersion` which happens for instance when you call `getReaderSchema`. In some sinks we call `getReaderchema` on each received message so it's done a lot of times.
There could be optimizations to prevent from decoding twice by optimizing `SchemaInfoProvider` but it's not the purpose of this PR.

**Comment by michaeljmarshall** [comment]
@cbornet - in that case, I don't think this PR should be labeled as fixing #15118. It seems you've identified two problems, and we should fix both.

One question about your caching solution. It seems your solution does not expire cache entries. I am not familiar with schema management, but wouldn't it make sense to use a caffeine cache to prevent unbounded growth on the `schemaMap` map?

**Review by eolivelli — APPROVED** [review]
LGTM

**Comment by cbornet** [comment]
> @cbornet - in that case, I don't think this PR should be labeled as fixing https://github.com/apache/pulsar/issues/15118. It seems you've identified two problems, and we should fix both.

Well the intent of #15118 was to describe the issue I'm solving here. Maybe another issue can be opened for the other one

> One question about your caching solution. It seems your solution does not expire cache entries. I am not familiar with schema management, but wouldn't it make sense to use a caffeine cache to prevent unbounded growth on the schemaMap map?

Yes the map is unbounded. I believe there won't be thousands of versions of a schema but I could be wrong.
There's already an unbounded cache of this type in AutoConsumeSchema

https://github.com/apache/pulsar/blob/193f5b2f74e919c05d0df651981cef439f55472f/pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoConsumeSchema.java#L51

**Review by michaeljmarshall — APPROVED** [review]
@cbornet - okay, thanks for clarifying about the cardinality of the cache. Please create the additional issue to track the double decoding of the schema.

It looks like a test is failing and it might be related. Please investigate.

**Review by lhotari — APPROVED** [review]

**Pull request merged**